src/wp-includes/media.php

```diff
@@ -5651,104 +5624,154 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 		return $loading_attrs;
 	}
 
-	if ( isset( $attr['loading'] ) ) {
-		/*
-		 * While any `loading` value could be set in `$loading_attrs`, for
-		 * consistency we only do it for `loading="lazy"` since that is the
-		 * only possible value that WordPress core would apply on its own.
-		 */
-		if ( 'lazy' === $attr['loading'] ) {
-			$loading_attrs['loading'] = 'lazy';
-			if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
-				_doing_it_wrong(
-					__FUNCTION__,
-					__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
-					'6.3.0'
-				);
+	/*
+	 * Skip programmatically created images within post content as they need to be handled together with the other
+	 * images within the post content.
+	 * Without this clause, they would already be considered within their own context which skews the image count and
+	 * can result in the first post content image being lazy-loaded or an image further down the page being marked as a
+	 * high priority.
+	 */
+	switch ( $context ) {
+		case 'the_post_thumbnail':
+		case 'wp_get_attachment_image':
+		case 'widget_media_image':
+			if ( doing_filter( 'the_content' ) ) {
+				return $loading_attrs;
 			}
-		}
-
-		return $postprocess( $loading_attrs, true );
-	}
-
-	// An image with `fetchpriority="high"` cannot be assigned `loading="lazy"` at the same time.
-	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
-		return $postprocess( $loading_attrs, true );
 	}
 
 	/*
-	 * Do not lazy-load images in the header block template part, as they are likely above the fold.
-	 * For classic themes, this is handled in the condition below using the 'get_header' action.
+	 * The key function logic starts here.
 	 */
-	$header_area = WP_TEMPLATE_PART_AREA_HEADER;
-	if ( "template_part_{$header_area}" === $context ) {
-		// Increase media count if there are images in header above a certian minimum size threshold.
-		$maybe_increase_content_media_count();
-		return $postprocess( $loading_attrs, true );
-	}
-
-	// The custom header image is always expected to be in the header.
-	if ( 'get_header_image_tag' === $context ) {
-		// Increase media count if there are images in header above a certian minimum size threshold.
-		$maybe_increase_content_media_count();
-		return $postprocess( $loading_attrs, true );
-	}
+	$in_viewport          = null;
+	$maybe_increase_count = false;
 
-	// Special handling for programmatically created image tags.
-	if ( 'the_post_thumbnail' === $context || 'wp_get_attachment_image' === $context || 'widget_media_image' === $context ) {
+	// Logic to handle a `loading` attribute that is already provided.
```

Comment:
I know we're generally trying to get rid of early returns with this refactor, but I wonder it it would be worth returning early if attributes already exist and we're not going to override them since the outcome is already known?

Reply:
I think it would be better not to, since even when the attribute is provided, we should still consider the element they are provided on in some of the logic below related to side effects. For example, images should still be counted based on the regular rules, or if an image has `fetchpriority="high"` on it already, it should trigger the logic to not apply the attribute to another later element.

That functionality was called in the original early returns too, so if we were to still return early we would need to keep the closures etc, so I think that's something to avoid. That's different from the early returns kept here which are for cases where we literally _disallow_ tweaking the attributes (e.g. when there's no width/height provided).